fix: enhance ZCL specification

Nerivec opened the pull request:
Introduces new typing for ZCL spec, with initial "auto update" using data from CSA ZAP (https://github.com/project-chip/zap)
Should eventually allow much better processing, e.g.:
- proper processing of non-values & default _per attribute_
- allow extending support (e.g. legacy devices missing a command parameter must rely on default to avoid failures)
- allow clamping numeric values as appropriate, but most importantly detect values out of range (which would at least be useful to log for debug purposes)
- allow better error handling (e.g. don't consider failure when not required)
- allow far better frontend validation & UX (prevent "doomed-to-fail" & co)

Notes:
- using incremental _per commit_ updates of Clusters for easier checking
- using a couple of "reports" for auto updates of Clusters for easier checking (present in same commit)

TODO:

- [x] update cluster typing
- [x] update clusters with ZAP data (auto)
- [ ] check clusters from auto update
- [ ] update clusters with spec data (manual, _for clusters not available with auto update_)
- [ ] create logic to handle default/invalid (will handle https://github.com/Koenkk/zigbee-herdsman/issues/1498)
- [ ] [optional/future] create logic to handle other metas
- [ ] [optional/future] add touchlink-specific types to parse from Buffalo
- [ ] move 3 sprut clusters to custom clusters (uses IDs not intended for custom)
- [ ] found spec typos (error-prone): `installedClosedLimitTiltDdegree`, `acCollTemp` (needs cross-repo check)
- [ ] remove `clusters-old.ts` reference (used by scripts for reporting)
- [ ] remove no longer useful scripts

CC: @sjorge @kirovilya

sjorge commented:
Big change, I had a quick look over it so far, I love the switch to constantly using hex makes things so much easier when reading the ZCL.

So looking at the scripts, is there an XML source of all the known clusters/attribute and they are now being parsed?

Nerivec commented:
Yes, it's in the ZAP repo I linked ([zap/zcl-builtin/dotdot](https://github.com/project-chip/zap/tree/master/zcl-builtin/dotdot)). Though it does not have _all_ spec clusters for some reason, so some will have to be done manually.

_ZAP could technically be used to generate the whole ZCL Clusters object for ZH, that's one of its purposes, but the problem is backwards compat with name changes (& customs), that would create a massive amount of changes needed in ZHC (& external converters). So, had to build a script to merge data instead._

Nerivec commented:
1213 pages later... 😅 

Important stuff to check:
- [changelog](https://github.com/Nerivec/zigbee-herdsman/blob/zap/scripts/clusters-changes.log) of existing data changes in `scripts/clusters-changes.log` (e.g. datatype changes, attribute additions, etc.)
  - see notes in first post (re. importance of some datatype changes)
  - _as a (much) shorter alternative to re-checking the whole spec, these highlight changes to prior behavior only (+ additions)_
- main logic changes in `zspec/zcl/utils.ts` (with specific tests added in `test/zspec/zcl/utils.test.ts`) & `zspec/zcl/buffaloZcl.ts` (mostly cleanup)
- `zspec/zcl/definition/datatypes.ts` is auto-generated from ZAP (we can deprecated some of the manually defined stuff eventually)
- keeping prior behavior:
  - types for Foundation are not defined specifically to uint, so moved everything to data (no non-values) - _more of a judgment guess really, didn't find an explicit reference_
  - custom parameters with non-values were marked with min/max to skip non-values (can be cleaned up later as needed)
    - e.g. `type: DataType.UINT8, max: 0xff`
  - custom attributes were all marked `writable` (except when comments explicitly specified otherwise) and with min/max to skip non-values (can be cleaned up later as needed)
    - e.g. `type: DataType.UINT8, writable: true, max: 0xff`
    - e.g. `type: DataType.INT8, writable: true, min: -128`

_Note: I skipped a few of the massive clusters that are not currently implemented and have no known use (added comments about them though)._

Koenkk commented on scripts/clusters-changes.log:
```diff
@@ -0,0 +1,585 @@
+Found 582 changes.
+
+[REMOVED] Clusters > pm1Measurement: 1068 (0x042c)
+[REMOVED] Clusters > pm10Measurement: 1069 (0x042d)
```
Do this one and the one above need to be added as custom clusters in zhc?

Nerivec commented:
Cleaned up a bit, figured extra bytes saved with shorter property names in `Clusters` was worth it since this is published over MQTT.
Also updated the script linked above to match (just `writable`=>`write`).
Removed `genLevelCtrlForLightning` in favor of always using `genLevelCtrl` (to have just 1 cluster with same ID). It's a bit less restrictive than `genLevelCtrlForLightning` in metadata, but ZHC can do the rest as needed (with so many off-spec devices, probably best to use the base version instead of the derived one anyway). _Added comments in-place for stuff that would be different._ Should avoid making a mess with cache, scenes & co.

sjorge commented:
A bit late but congrats on landing this! <3
